README.md
computer-stats - a computer metric aggregation script

## Usage

computer-stats.pl [-o output-file] [-d disk mount]

Note: The -d flag can be used any number of times to specify additional directories.
	It should also be noted that the measurement takes place for the mount point
	of the disk used at the directory specified.
Example: A single disk, single partition install will have all -d paths give disk
	stats for /.
	Other configurations will use the partition mount point for the
	specified directory.

## Requirements

1. Perl 5.14 or newer.
2. A Unix-like system. Possibly specifically Linux, but more testing is needed to verify.

## Installation

**Linux**

Run install.sh, providing any arguments for the cron job to the script.
install.sh will need superuser privileges to operate correctly.

Example: `sudo install.sh -o /var/log/output.csv -d /home -d /` will make the logfile be written
at `/var/log/output.csv` and the program will check the disk space of the partitions at `/home` and `/`.

**Windows**

There is no Windows-compatible version for this script at this time.

**Mac OS X**

Memory information fails due to a lack of /proc/meminfo. Not compatible without modification.

**BSD**

Try Linux instructions above. Not tested whether this will work or not.

## Build Status

[![tv-image][]][tv-site]

[tv-image]: https://travis-ci.org/SilverNexus/computer-stats.svg?branch=master
[tv-site]: https://travis-ci.org/SilverNexus/computer-stats/branches

## License

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


ChangeLog
2018-02-22  Daniel Hawkins
    * computer_stats.pl: Change day separator in uptime to ; to prevent conflict
      with the field delimiter.

2017-06-22	Daniel Hawkins
	* computer_stats.pl: Fix improper regex handling of (1 <= day < 2) uptime.
	  Fix improper regex handling of n days, x min uptime.

2016-08-30	Daniel Hawkins
	* computer stats.pl: Remove development output from script.
	* install.sh: Fix typo in comments.
	  Fix missing asterisk in installed crontab line.
	* README.md: Update installation instructions to use new install script.
	* .travis.yml: Update travis-ci tests to include test for the install script.
	* TODO: Obsoleted the TODO file. Not sure why I didn't just use the issue
	  tracker to begin with.

2016-08-25	Daniel Hawkins
	* computer-stats.pl: Added feature for multiple disks and cleaned up argument handling.
	* README.md: Updated usage information to reflect new argument handling.
	* .travis.yml: Updated existing checks to reflect new argument handling.
	* install.sh: Added install script.

2016-08-23	Daniel Hawkins
	* computer-stats.pl: Add the ability to set the output file name and location.
	* README.md: Update to reflect new command-line argument.
	* .travis.yml: Add test to ensure that the use of a custom output file works.

2016-08-21	Daniel Hawkins
	* computer-stats.pl: Initial implementation.
	* .travis.yml: Build testing automation. Includes tests for Mac OS X that fail.
	* COPYING: Issued under GNU GPLv2
	* README.md: Established minimum requirements and installation instructions.
	* TODO: List of future plans for the tool.


computer-stats.pl
#!/usr/bin/env perl

#    computer_stats.pl - usage aggregation tool for the computer. Best used as a cron job.
#    Copyright (C) 2016  SilverNexus
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

use strict;
use Getopt::Long;

my $filename = "usage_log.csv";
# Reference to a scalar.
my @disk_path_list;

# Use Getopt to clean up argument parsing
GetOptions("output-file|o=s" => \$filename,
	   "disk-paths|d=s" => \@disk_path_list);

# Get the memory and swap information
my $res = `egrep "Mem|Cache|Swap|Buffer" /proc/meminfo` or die("Could not access memory information.");
# Get the workload information
my $uptime = `uptime` or die("Could not access workload information.");

# Get the time, too
my $time = `date` or die("Could not access time.");
chomp($time);
# And get a list of users
my $userlist = `users` or die("Could not retrieve user list.");
chomp($userlist);

# Get the specified disks' usage
my $disk = `df @disk_path_list` or die("Could not check disk usage.");
my @lines = split("\n", $disk);
# The first line is just titles -- skip it
shift(@lines);

# Split the memory results by line
# We use multiple lines of this, so we do this after the other commands to allow reuse of variables.
my @memlines = split("\n", $res);

#Declare the variables to store the memory information
my $total, my $free, my $swap, my $swfree, my $cache, my $buffers, my $up,
my $load1, my $load5, my$load15, my $numusers, my @disksize, my @diskfree,
my @diskused, my @diskmount, my @diskpct;

#Split out the memory and swap information
for(@memlines){
    if (/^MemTotal:\s*(\d+)/s){
        $total = $1;
    }
    elsif (/^MemFree:\s*(\d+)/s){
        $free = $1;
    }
    elsif (/^SwapTotal:\s*(\d+)/s){
        $swap = $1;
    }
    elsif (/^SwapFree:\s*(\d+)/s){
        $swfree = $1;
    }
    elsif (/^Cached:\s*(\d+)/s){
        $cache = $1;
    }
    elsif (/^Buffers:\s*(\d+)/s){
        $buffers = $1;
    }
}

# Split out the uptime information.
# For the uptime, it needs to handle multiple days of uptime, as well as under an hour.
if ($uptime =~ /\s*\d{2}:\d{2}:\d{2}\s+up\s+((\d+ days?,\s+)?((\d+:\d{2})|(\d+ min))),\s+(\d+) users?,\s+load average:\s+(\d+\.\d{2}),\s+(\d+\.\d{2}),\s+(\d+\.\d{2})/s){
    $up = $1;
    $numusers = $6;
    $load1 = $7;
    $load5 = $8;
    $load15 = $9;
    # Convert the comma from more than one day of uptime to avoid conflicts with the line delimiter.
    $up =~ s/\,/\;/g;
}

# Split out disk information
for (@lines){
    # TODO: Come up with a better way to handle the multiple disks -- current way is ugly.
    if (/^((\/\w+)+)\s*(\d+[GMKT]?)\s*(\d+[GMKT]?)\s*(\d+[GMKT]?)\s*(\d+\%)\s*(\/(\w+\/?)*)/s){
	push(@diskmount, $7);
	push(@disksize, $3);
	push(@diskused, $4);
	push(@diskfree, $5);
	push(@diskpct, $6);
    }
}

# See if the file exists. If not, initialize the new log file with header information
initialize() unless (-e $filename);

# open the output file and append to it.
open(LOG, ">> $filename") or die("Failed to open $filename");
# Separate multiple disks by semicolon for better parsing.
$" = ";";
say LOG "$time, $up, $total, $free, $cache, $buffers, $swap, $swfree, @diskmount, @disksize, @diskused, @diskfree, @diskpct, $load1, $load5, $load15, $numusers, $userlist";
close(LOG);

sub initialize{
    open(LOGINIT, "> $filename") or die("Failed to initialize $filename");
    say LOGINIT "Time, Uptime, RAM Total (KiB), RAM Free (KiB), Cache (KiB), Buffers (KiB), Swap Total (KiB), Swap Free (KiB), Disk Mount Point, Disk Size (KiB), Disk Used (KiB), Disk Available (KiB), Disk Used (%), CPU Load (1 min), CPU Load (5 min), CPU Load (15 min), # Users, Users";
    close(LOGINIT);
}


install.sh
#!/bin/sh

# install.sh -- Install script for computer-stats
#    Copyright (C) 2016  SilverNexus
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

if [ $USER != "root" ];
then
    echo "You need to run this script as root to install."
    exit 1
fi

line="*/10 * * * root perl /etc/cron.d/computer-stats.pl $*"

cp computer-stats.pl /etc/cron.d/

# Make sure we dont already have a line in the crontab. If we do, remove it.
res=`grep -e "computer-stats.pl" /etc/crontab`

if [ "$res" != "" ];
then
    # Replace that line in crontab with a new one.
    # Writing directly to crontab creates an empty file, so we make a new file,
    # then turn around and copy it back to crontab
    sed "/computer-stats.pl/ c $line" /etc/crontab > /etc/crontab_new
    mv /etc/crontab_new /etc/crontab
    touch /etc/crontab
else
    # Just append it to the crontab
    echo "$line" >> /etc/crontab
fi
